### PR TITLE
[DNM] common/msg: sockaddr on FreeBSD differs from Linux, has sa_len

### PR DESCRIPTION
**Pull request by wjwithagen**
On Linux the definition of a sockaddr is:
```
struct sockaddr {
        sa_family_t sa_family;
        char        sa_data[14];
}
typedef unsigned short __kernel_sa_family_t;
```
Short is atleast 16 bits.

And on FreeBSD ik looks like:
```
struct sockaddr {
	unsigned char	sa_len;		/* total length */
	sa_family_t	sa_family;	/* address family */
	char		sa_data[14];	/* actually longer; address value */
};
typedef       __sa_family_t   sa_family_t;
:typedef       __uint8_t       __sa_family_t;
```
So the `sa_data[]` array always starts after 2 bytes. 
But depending on architecture lands the LSB of a short in the first of the second byte of the structure.
On i386/amd64 it lands in the first byte where FreeBSD would expect sa_len.

We need to do:
 - maintain the sa_len value for FreeBSD, otherwise memcmp() calls
   can return unequal, where the only difference would be `sa_len`.
   sa_len is set on returns from the FreeBSD syscalls.
 - add getter and setter functions for sa_len and sa_family.
 - Look into comparing structures whch have sockaddr as members.

Testing with `ceph_test_msgr` will show all poll tests failing.
Most of the time the script wil just loop endlessly.

Future work could be to actually rework the comparission routines in
msg_types.h to make full use of the structure. But for now it looks
like it not needed.

Backport: Mimic, Luminous

Tracker: http://tracker.ceph.com/issues/38457

- [X] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by wjwithagen** [comment]
@smithfarm 
I tried making a tracker with backport information.
http://tracker.ceph.com/issues/38457
But there is no place to put any backport info in a new issue???
So where do I put that.

File: src/msg/msg_types.h [hunk 1]
```diff
@@ -462,6 +475,10 @@ struct entity_addr_t {
   // Right now, these only deal with sockaddr_storage that have only family and content.
   // Apparently on BSD there is also an ss_len that we need to handle; this requires
   // broader study
+  // Currently ss_len is set on the decode in set_sockaddr() so it does not need to
+  // be encode to go over the line.
+  // But get_sockaddr_storage() does include ss_len, so it could be that line formats
+  // between Linux and BSD are not compatible
```
Comment (liewegas) [inline src/msg/msg_types.h]: note that with v2 we aren't send a whole sockaddr_storage over the wire.  does the newer encoding make this issue go away?

File: src/msg/simple/Pipe.cc [hunk 2]
```diff
@@ -1115,7 +1115,7 @@ int Pipe::connect()
   port = peer_addr_for_me.get_port();
 
   ldout(msgr->cct,20) << "connect read peer addr " << paddr << " on socket " << sd << dendl;
-  if (peer_addr != paddr) {
+  if (!peer_addr.is_same_host(paddr)) {
```
Comment (liewegas) [inline src/msg/simple/Pipe.cc]: is this change needed with the other adjustments?

File: src/msg/msg_types.h (hunk 1 above)
Comment (wjwithagen) [reply src/msg/msg_types.h]: @liewegas 
I noticed that, but for v1 it is still the case.
So perhaps add that restriction to the text. 
Haven't had time to test interoperability for quite some time, so I could very well be that this is still relevant. Even more with all the v2 changes, and this becoming less obvious.

File: src/msg/simple/Pipe.cc (hunk 2 above)
Comment (wjwithagen) [reply src/msg/simple/Pipe.cc]: @liewegas 
I was actually going to rewrite all the compare tests, to exclude any dependancies on `{sa,ss}_len` but that turns out to be relatively messy. :( So this was the next best thing.
I'll rerun the tests with `!=` there, and amend if it runs oke with `cep_test_msgr`

**Comment by wjwithagen** [comment]
@tchaikov 
I noticed I did not include you on the review, but could you please?

**Comment by wjwithagen** [comment]
@liewegas @rjfd @tchaikov 
Reworked some of the code and set it to DNM until reviews are done.
So please fire at will.

File: src/msg/msg_types.h [hunk 3]
```diff
@@ -175,21 +179,34 @@ WRITE_CLASS_ENCODER(ceph_sockaddr_storage)
...
   ceph_sockaddr_storage ss{};
   auto src = (unsigned char const *)&a;
   auto dst = (unsigned char *)&ss;
   src += sizeof(a.ss_len);
   ss.ss_family = a.ss_family;
   src += sizeof(a.ss_family);
   dst += sizeof(ss.ss_family);
-  const auto copy_size = std::min((unsigned char*)(&a + 1) - src,
-				  (unsigned char*)(&ss + 1) - dst);
+  // const auto copy_size = std::min((unsigned char*)(&a + 1) - src,
```
Comment (tchaikov) [inline src/msg/msg_types.h]: why the existing `copy_size` is wrong?

File: src/msg/msg_types.h [hunk 4]
```diff
@@ -175,21 +179,34 @@ WRITE_CLASS_ENCODER(ceph_sockaddr_storage)
...
+  //				  (unsigned char*)(&ss + 1) - dst);
+  const auto copy_size = a.ss_len - sizeof(a.ss_family) - sizeof(a.ss_len);
```
Comment (tchaikov) [inline src/msg/msg_types.h]: i think we should also memcpy the padding bytes to make sure all bits in `ss` are properly set. after this change, the behavior is broken.

File: src/msg/msg_types.h [hunk 5]
```diff
@@ -17,6 +17,10 @@
 
 #include <sstream>
 
+// #include <string>
```
Comment (tchaikov) [inline src/msg/msg_types.h]: please remove these `#include` if they are not related to this change.

File: src/msg/msg_types.cc [hunk 6]
```diff
@@ -129,9 +129,15 @@ bool entity_addr_t::parse(const char *s, const char **end, int default_type)
   if (inet_pton(AF_INET, buf4, &a4)) {
     u.sin.sin_addr.s_addr = a4.s_addr;
     u.sa.sa_family = AF_INET;
+#if defined(__FreeBSD__)
+    u.sa.sa_len = sizeof(u.sin);
+#endif
     p = start + strlen(buf4);
   } else if (inet_pton(AF_INET6, buf6, &a6)) {
     u.sa.sa_family = AF_INET6;
+#if defined(__FreeBSD__)
```
Comment (tchaikov) [inline src/msg/msg_types.cc]: instead of hardwiring this to `defined(__FreeBSD__)`. i'd suggest check the `sa_len` using

```cmake
check_struct_has_member("struct sockaddr" sa_len sys/socket.h HAVE_STRUCT_SOCKADDR_SA_LEN LANGUAGE C)
```

File: src/msg/msg_types.h [hunk 7]
```diff
@@ -350,9 +368,12 @@ struct entity_addr_t {
     u.sin.sin_family = AF_INET;
     unsigned char *ipq = (unsigned char*)&u.sin.sin_addr.s_addr;
     ipq[pos] = val;
+#if defined(__FreeBSD__)
+    u.sa.sa_len = get_sockaddr_len();
+#endif
   }
   void set_port(int port) {
-    switch (u.sa.sa_family) {
+    switch (get_family()) {
```
Comment (tchaikov) [inline src/msg/msg_types.h]: please move this change into its own commit.

File: src/msg/msg_types.h [hunk 8]
```diff
@@ -393,7 +416,7 @@ struct entity_addr_t {
       return false;
     if (is_blank_ip() || o.is_blank_ip())
       return true;
-    if (memcmp(&u, &o.u, sizeof(u)) == 0)
+    if (is_same_host(o))
```
Comment (tchaikov) [inline src/msg/msg_types.h]: i think, after your change `memcpy()` should serve the purpose just well?

File: src/msg/msg_types.h (hunk 3 above)
Comment (wjwithagen) [reply src/msg/msg_types.h]: @tchaikov 
Because it depends on how the variables are layout out on the stack?
Besides the fact that comparing pointers is sort of trickery, which I usually don't really like since it is hard to follow. What I'm suggesting is a more logic evaluation of what is going on.

File: src/msg/msg_types.h (hunk 4 above)
Comment (wjwithagen) [reply src/msg/msg_types.h]: @tchaikov 
I guess we could, since in both platform the max size of the sockaddr (storage) is like 128 bytes.
On the otherhand we could init the block first with 0's??

But is the full sockaddr_storage send on the wire? What kind of information is there that we should be copying?

File: src/msg/msg_types.h (hunk 5 above)
Comment (wjwithagen) [reply src/msg/msg_types.h]: @tchaikov 
Done that already in local commits, they will be rebase in next commits

File: src/msg/msg_types.cc (hunk 6 above)
Comment (wjwithagen) [reply src/msg/msg_types.cc]: @tchaikov 
Ah, new Cmake trick.... Will try to work that out.

File: src/msg/msg_types.h (hunk 8 above)
Comment (wjwithagen) [reply src/msg/msg_types.h]: @tchaikov 
Until things change again, especially with we compare structs where it could be unclear what is actually stored after the sa_address/sa_data field. Even more so when we start comparing structs that originate from Linux and FreeBSD on a hybrid cluster. 
So for most purposes I would like te get ride of all memcmp's here, but the `<, >, => =<` ones are a bit harder to get the logic right.

File: src/msg/msg_types.h (hunk 3 above)
Comment (tchaikov) [reply src/msg/msg_types.h]: > Because it depends on how the variables are layout out on the stack? 

this does not render it *wrong*.

File: src/msg/msg_types.h (hunk 4 above)
Comment (tchaikov) [reply src/msg/msg_types.h]: for v1, yes. and per Sage, we are not sending the whole `sockaddr_storage ` over the wire. i think he replied you in the same PR.

File: src/msg/msg_types.h (hunk 8 above)
Comment (tchaikov) [reply src/msg/msg_types.h]: on the contrary, i prefer `memcmp()` over the hand-written `<`. because every time we send a message, we look up the connection by the peer address. so it's sitting in the critical path, and we need to make it as fast as possible.

File: src/msg/msg_types.h (hunk 4 above)
Comment (wjwithagen) [reply src/msg/msg_types.h]: @tchaikov 
There is a wierd catch in this code, since Linux does not have sa_len, it does not suffer from that fact that a ANY_ADDRESS (aka all 0's) suffers from the fact that the length of the structure and sa_len are not equal. 
When sa_len exists it is 0 when the address is ANY_ADDRESS, so I'm splitting the size calculation of the structure from sa_len calculation, since they differ exactly on this topic.
Resulting in crashes from allocating `0xffffffff` sized buffers.

**Comment by tchaikov** [comment]
@wjwithagen are you still working on this change? is it ready for another round of review?

**Comment by wjwithagen** [comment]
@tchaikov 
Sort of .....
At work seriously occupied with getting ISO reviews and certifiations done. Whihc has kept me from working on Ceph for the last few weeks.
And first need to fix adding Boost 1.7 on FreeBSD, and then figure out why the python ceph program refuses to connect to the admin socket.

**Comment by wjwithagen** [comment]
Revive, still need to figure this one out.
But the major errors it caused have gone for quite some time.

**Comment by wjwithagen** [comment]
ping

**Comment by wjwithagen** [comment]
@tchaikov 
I'mm going to rework this PR to get this `sa_len` thing out of my system.
It is causing trouble in `unittest_compression` as well
    https://tracker.ceph.com/issues/45390

**Comment by wjwithagen** [comment]
Still working on this

**Comment by wjwithagen** [comment]
Ping alive

**Comment by wjwithagen** [comment]
Ping alive
